packages/daemon/src/daemon.js

```diff
@@ -775,19 +803,60 @@ const makeDaemonCore = async (
   };
 
   /**
-   * @param {string} workerFormulaIdentifier
+   * @param {string} hostFormulaIdentifier
    * @param {string} source
    * @param {string[]} codeNames
-   * @param {string[]} endowmentFormulaIdentifiers
+   * @param {(string | string[])[]} endowmentFormulaPointers
+   * @param {import('./types.js').EvalFormulaHook[]} hooks
+   * @param {string} [specifiedWorkerFormulaIdentifier]
    * @returns {Promise<{ formulaIdentifier: string, value: unknown }>}
    */
   const incarnateEval = async (
-    workerFormulaIdentifier,
+    hostFormulaIdentifier,
     source,
     codeNames,
-    endowmentFormulaIdentifiers,
+    endowmentFormulaPointers,
+    hooks,
+    specifiedWorkerFormulaIdentifier,
   ) => {
-    const formulaNumber = await randomHex512();
+    const {
+      workerFormulaIdentifier,
+      endowmentFormulaIdentifiers,
+      evalFormulaNumber,
+    } = await formulaGraphMutex.enqueue(async () => {
+      const ownFormulaNumber = await randomHex512();
+      const workerFormulaNumber = await (specifiedWorkerFormulaIdentifier ??
+        randomHex512());
+
+      const identifiers = harden({
+        workerFormulaIdentifier: (
+          await incarnateWorkerSync(workerFormulaNumber)
+        ).formulaIdentifier,
+        endowmentFormulaIdentifiers: await Promise.all(
+          endowmentFormulaPointers.map(async formulaIdOrPath => {
```

Comment:
I think `formulaIdOrPath` is a misnomer here. I recommend just using the name `petNamePath` and just fall through to `petName` if it’s a string. It should never be a nonce / formula identifier.

Reply:
See: https://github.com/endojs/endo/pull/2092/files#r1498644209